File: 2975-maximum-square-area-by-removing-fences-from-a-field/2975-maximum-square-area-by-removing-fences-from-a-field.py
class Solution:
    def maximizeSquareArea(self, m: int, n: int, h_fences: List[int], v_fences: List[int]) -> int:
        h = sorted(h_fences + [1, m])
        v = sorted(v_fences + [1, n])
        dh = {h[i] - h[j] for i in range(len(h) - 1, 0, -1) for j in range(i)}
        dv = [v[i] - v[j] for i in range(len(v) - 1, 0, -1) for j in range(i)]
        ans = next((x for x in sorted(dv, reverse=True) if x in dh), -1)
        return ans ** 2 % (10**9 + 7) if ans != -1 else -1

File: 2975-maximum-square-area-by-removing-fences-from-a-field/2975-maximum-square-area-by-removing-fences-from-a-field.java
class Solution {
    public int maximizeSquareArea(int m, int n, int[] hFences, int[] vFences) {

    Arrays.sort(hFences); //mlogn
    Arrays.sort(vFences); //nlogn

	List<Integer> hFencesList = new ArrayList<>();//m space
	for(int h : hFences) hFencesList.add(h);		
      hFencesList.add(0,1);
      hFencesList.add(m);

    List<Integer> vFencesList = new ArrayList<>(); //n space
	  for(int v : vFences) vFencesList.add(v);		
      vFencesList.add(0,1);
      vFencesList.add(n);

	Set<Integer> matcherSet = new HashSet<>(); //m^2
	for(int i=0;i<hFencesList.size();i++) {//m^2
		for(int j=i+1;j<hFencesList.size();j++) {
			matcherSet.add(Math.abs(hFencesList.get(j)-hFencesList.get(i)));
		}
    }
	
    long result = -1;
	for(int i=0;i<vFencesList.size();i++) { //n^2
		for(int j=i+1;j<vFencesList.size();j++) {
			int diff = vFencesList.get(j)-vFencesList.get(i);
			if(matcherSet.contains(diff)){
				result = Math.max(result,diff); 
            }
		}
    }

    return result == -1 ? -1 : (int) ((result * result) % (1000000007));
    }
  }